### DISCUSSION TITLE
Typescript transformer support for paginated collection

### DISCUSSION THREAD

**Discussion by Wulfheart** [spatie/laravel-data#131] [Ideas]
As far as I can see when transforming to typescript the (paginated) collection is not generated. Is this planned for v2? @rubenvanassche (tagging you because I think you are the main guy for v2)

**Comment by rubenvanassche** [comment]
Not at the moment, because we're not using typed paginated collections. That being said, I would accept a PR adding such feature since it probably can be quite useful for everyone!

**Reply by Wulfheart** [reply]
What do you use instead of paginated collections?

**Reply by rubenvanassche** [reply]
We do use them, but don't automatically generate them. Instead a type like this is used:

```tsx
type PaginatedCollection<T extends object> = {
    data: Array<T>;
    meta: {
        current_page: number;
        first_page_url: string | null;
        from: number;
        last_page: number;
        last_page_url: string | null;
        next_page_url: string | null;
        path: string;
        per_page: string;
        prev_page_url: string | null;
        to: number;
        total: number;
    };
};

```

**Reply by UltraWelfare** [reply]
Is this still the case in v4?

**Reply by rubenvanassche** [reply]
This is planned for TS transformer 3

**Reply by Bosphoramus** [reply]
Hi @rubenvanassche 👋🏻
Would love to help with this, is there any initial work for pagination objects TS transform on the [`v3` branch of typescript-transformer](https://github.com/spatie/typescript-transformer/tree/v3)?

**Comment by DevBertl** [comment]
For Laravel 12:

- If you use simplePaginate():

```
type SimplePaginatedCollection<T extends object> = {
    current_page: number;
    current_page_url: string;
    data: Array<T>;
    first_page_url: string;
    from: number;
    next_page_url: String;
    path: String;
    per_page: number;
    prev_page_url: String;
    to: number;
}
```

Use it like so:

`export interface SimpleCustomers extends SimplePaginatedCollection<Customer> {}`


- If you use paginate():

```
type PaginatedCollection<T extends object> = {
    current_page: number;
    data: Array<T>;
    first_page_url: String;
    from: number;
    last_page: number;
    last_page_url: String;
    links: Array;
    next_page_url: String;
    path: String;
    per_page: number;
    prev_page_url: String;
    to: number;
    total: number;
};
```

Use it like so:

`export interface Customers extends PaginatedCollection<Customer> {}`
